ledger/migration/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Ledger migration: Perform updates of LedgerDB to accommodate for
+//! backward-incompatible changes.
+
+use lmdb::{DatabaseFlags, Environment, Transaction, WriteFlags};
+use mc_common::logger::{log, Logger};
+use mc_ledger_db::{
+    key_bytes_to_u64, tx_out_store::TX_OUT_INDEX_BY_PUBLIC_KEY_DB_NAME, u64_to_key_bytes, Error,
+    LedgerDbMetadataStoreSettings, MetadataStore, TxOutStore, TxOutsByBlockValue,
+    BLOCK_NUMBER_BY_TX_OUT_INDEX, COUNTS_DB_NAME, NUM_BLOCKS_KEY, TX_OUTS_BY_BLOCK_DB_NAME,
+};
+use mc_util_lmdb::MetadataStoreError;
+use mc_util_serial::decode;
+use std::path::Path;
+
+const MAX_LMDB_FILE_SIZE: usize = 1_099_511_627_776; // 1 TB
+
+pub fn migrate(ledger_db_path: impl AsRef<Path>, logger: &Logger) {
+    // Open the LMDB database.
+    let env = Environment::new()
+        .set_max_dbs(24)
+        .set_map_size(MAX_LMDB_FILE_SIZE)
+        .open(ledger_db_path.as_ref())
+        .expect("Failed opening ledger db");
+
+    // Create metadata store.
+    let metadata_store = MetadataStore::<LedgerDbMetadataStoreSettings>::new(&env)
+        .expect("Failed creating MetadataStore");
+
+    // Incrementally perform upgrades until we reach the current version.
+    loop {
+        // Check if the database we opened is compatible with the current
+        // implementation.
+        let db_txn = env.begin_ro_txn().expect("Failed starting ro transaction");
+        let version = metadata_store
+            .get_version(&db_txn)
+            .expect("Failed getting metadata version");
+        log::info!(logger, "Ledger db is currently at version: {:?}", version);
+        db_txn.commit().expect("Failed committing transaction");
+
+        match version.is_compatible_with_latest() {
+            Ok(_) => {
+                break;
+            }
+            // Version 20200610 introduced the TxOut public key -> index store.
+            Err(MetadataStoreError::VersionIncompatible(20200427, 20200610))
+            | Err(MetadataStoreError::VersionIncompatible(20200427, 20200707)) => {
```

Comment:
Consider inferring the second parameter so we don't have to list all later versions.

```suggestion
            Err(MetadataStoreError::VersionIncompatible(20200427, _)) => {
```

Reply:
good catch, I'll also improve the comments around this a bit.